test/cli/content_routing_http_test.go

```diff
@@ -0,0 +1,115 @@
+}
+
+func (r *mockHTTPContentRouter) ProvideBitswap(ctx context.Context, req *server.BitswapWriteProvideRequest) (time.Duration, error) {
+	args := r.Called(ctx, req)
+	return args.Get(0).(time.Duration), args.Error(1)
+}
+func (r *mockHTTPContentRouter) Provide(ctx context.Context, req *server.WriteProvideRequest) (types.ProviderResponse, error) {
+	args := r.Called(ctx, req)
+	return args.Get(0).(types.ProviderResponse), args.Error(1)
+}
+
+// userAgentRecorder records the user agent of every HTTP request
+type userAgentRecorder struct {
+	delegate   http.Handler
+	userAgents []string
+}
+
+func (r *userAgentRecorder) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	r.userAgents = append(r.userAgents, req.UserAgent())
+	r.delegate.ServeHTTP(w, req)
+}
+
+func TestContentRoutingHTTP(t *testing.T) {
+	cr := &mockHTTPContentRouter{}
+	cr.Test(t)
+
+	// run the content routing HTTP server
+	listener, err := net.Listen("tcp", "127.0.0.1:")
+	require.NoError(t, err)
+	t.Cleanup(func() { listener.Close() })
+	userAgentRecorder := &userAgentRecorder{delegate: server.Handler(cr)}
+	go func() {
+		err := http.Serve(listener, userAgentRecorder)
+		if err != nil && !errors.Is(err, http.ErrServerClosed) && !errors.Is(err, net.ErrClosed) {
+			t.Logf("HTTP server closed with error: %s", err)
+		}
+	}()
+
+	// setup the node
+	node := harness.NewT(t).NewNode().Init()
+	node.Runner.Env["IPFS_HTTP_ROUTERS"] = "http://" + listener.Addr().String()
+	node.StartDaemon()
+
+	// compute a random CID
+	randStr := string(testutils.RandomBytes(100))
+	res := node.PipeStrToIPFS(randStr, "add", "-qn")
+	wantCIDStr := res.Stdout.Trimmed()
+	wantCID, err := cid.Decode(wantCIDStr)
+	require.NoError(t, err)
+
+	cr.On("FindProviders", mock.Anything, wantCID).Return([]types.ProviderResponse{}, nil)
+
+	t.Run("fetching an uncached block results in an HTTP lookup", func(t *testing.T) {
+		statRes := node.Runner.Run(harness.RunRequest{
+			Path:    node.IPFSBin,
+			Args:    []string{"block", "stat", wantCIDStr},
+			RunFunc: (*exec.Cmd).Start,
+		})
+		defer func() {
+			if err := statRes.Cmd.Process.Kill(); err != nil {
+				t.Logf("error killing 'block stat' cmd: %s", err)
+			}
+		}()
+
+		// verify the content router was called
+		assert.Eventually(t, func() bool {
+			return len(cr.Calls) > 0
+		}, 10*time.Second, 10*time.Millisecond)
+
+		assert.NotEmpty(t, userAgentRecorder.userAgents)
+		version := node.IPFS("id", "-f", "<aver>").Stdout.Trimmed()
+		for _, userAgent := range userAgentRecorder.userAgents {
+			assert.Equal(t, version, userAgent)
+		}
+	})
+
+	// cid.contact supports GET-only: https://github.com/ipfs/kubo/issues/9504
+	// which means no announcements over HTTP should be made.
+	t.Run("adding a new CID does not result in an HTTP provide", func(t *testing.T) {
+		cr.Calls = nil
+		cr.ExpectedCalls = nil
+
+		randStr := string(testutils.RandomBytes(100))
+		node.IPFSAddStr(randStr)
+
+		assert.Zero(t, cr.Calls, "expected no HTTP router calls")
```

Comment:
Do we synchronously announce CIDs we add ?

I would not be surprised if this test pass even if we send a PUT, just because we run this check before the background announcing would lose the race to here.

Reply:
I'm pretty sure they are synchronously provided.

Reply:
I've added:
```diff
 func (r *fakeHTTPContentRouter) ProvideBitswap(ctx context.Context, req *server.BitswapWriteProvideRequest) (time.Duration, error) {
 	r.m.Lock()
 	defer r.m.Unlock()
+	time.Sleep(time.Second * 30)
 	r.provideCalls++
 	return 0, nil
 }
 func (r *fakeHTTPContentRouter) Provide(ctx context.Context, req *server.WriteProvideRequest) (types.ProviderResponse, error) {
 	r.m.Lock()
 	defer r.m.Unlock()
+	time.Sleep(time.Second * 30)
 	r.provideCalls++
 	return nil, nil
 }
```
And:
```console
$ go test -run=TestContentRoutingHTTP ./test/cli/
ok  	github.com/ipfs/kubo/test/cli	1.898s
```
I think something is dodgy here.

Reply:
I edited `go-libp2p-kad-dht`'s `Provide` with a `time.Sleep(30 * time.Second)` and `ipfs add` a single block file completes in `0.041ms`.

I don't belive this is true.
`ipfs add` would also needs to be much much slower with the non accelerated DHT client.

Reply:
(this is bad because this means this test either tests nothing or is flaky)

Reply:
I read through the code and it looked like it's happening synchronously. Can you show me the code where it's not?

Reply:
@hacdias we talked about this yesterday, this test is wrong.
This is not happening synchronously here: https://github.com/ipfs/boxo/blob/a3f1b66d8ef3ea00d139c96908bbf0ffb871ce7b/provider/batched/system.go#L338

This will write to the datastore synchronously, but what is done synchronously is the enqueing, the provide is then done later when the worker gets around to it.

Reply:
Oh this was removed, I didn't saw that, I guess that fine but that means we are not testing for this.

Reply:
Can't have your cake and eat it too, it's inherently not a testable design.